### DISCUSSION TITLE
Examples

### DISCUSSION THREAD

**Discussion by MAKOMO** [pymodbus-dev/pymodbus#1815] [Help] [answered]
I am trying to run the Advanced Examples from here

https://pymodbus.readthedocs.io/en/latest/source/examples.html#

The linked archive [examples.zip](https://pymodbus.readthedocs.io/en/latest/_downloads/5fcb20102e97ee54994f817e872e8e91/examples.zip) seems to be empty while the linked [examples.tgz](https://pymodbus.readthedocs.io/en/latest/_downloads/82fca487250825202900cae86f76d7ea/examples.tgz) seems to contain the files.

It is adviced to run the examples from within the `examples` folder, but this does not work for me (macOS/Python.org v3.11).

```
[Max:/Users/luther/Downloads/examples] ./server_async.py
Traceback (most recent call last):
  File "/Users/luther/Downloads/examples/./server_async.py", line 38, in <module>
    from examples import helper
ModuleNotFoundError: No module named 'examples'
```

**Comment by janiversen** [comment]
it's a bug that the zip file is empty.

Last time I tried with the tgz file it worked, but let me check it again.

**Comment by janiversen** [comment]
"latest" do have a problem ! Funny thing is that when I downloaded from "dev" it worked, but now I only get 168bytes...it is a download problem, because the release contains the right size for both zip and tgz

The problem with "examples" module missing, might be because we test everything within a cloned directory.

Anyhow thanks for reporting the problems I will have a look later to see what can be done.

**Reply by MAKOMO** [reply]
Thanks for looking into this. No hurry!

I am trying to follow your advice to replace the (serial) sync clients by async clients as they are way faster in one of the recent issues. I did struggle to get the typing right for this in my code and thus looked at your examples. While the `client_performance.py` example seems to work as is and I also can duplicate the noted performances advantage, mypy complains also on this code

```
client_performance.py:67: error: Incompatible types in "await" (actual type "ModbusResponse", expected type "Awaitable[Any]")  [misc]
```

Removing the `await` in 
```
rr = await client.read_input_registers(980, REGISTER_COUNT, slave=1)
```
like in
```
rr = client.read_input_registers(980, REGISTER_COUNT, slave=1)
```
makes mypy happy, but breaks the example.

The reason seems to be that the method `read_input_registers` is inherited from `ModbusClientMixin` where it is not declared `async def` but just `def`. Not sure why that works at all and how to tell mypy.

While running mypy (under its default settings) on `client_performance.py` it also complains a lot on "Implicit Optionals".

```
[Max:/Users/luther/Downloads/pymodbus-dev/examples] mypy client_performance.py
/Users/luther/Downloads/pymodbus-dev/pymodbus/logging.py:20: error: Incompatible default for argument "log_file_name" (default has type "None", argument has type "str")  [assignment]
/Users/luther/Downloads/pymodbus-dev/pymodbus/logging.py:20: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
/Users/luther/Downloads/pymodbus-dev/pymodbus/logging.py:20: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
```

I did run that `no_implicit_optional` thing using the flag `--use-union-or` to keep this compatible with Python 3.8 (making the tool to insert `Optional[X]` instead the somewhat nicer `X | None` which requires Python 3.10. I think it is a good idea to avoid those Implicit Optionals to get informed on wrong use of those types by mypy. Should I raise a PR on this or do you prefer to keep the code cleaner and keep using those `Implicit Optionals`?


One tiny thing more: the main page on GitHub reads "`Just fork the repo and raise your PR against [dev]{.title-ref} branch.`. Maybe just a typo. Anyhow, it seems you are revising those .rst files anyhow to .md, right?

**Reply by janiversen** [reply]
You need the "await" as I told you before it is perfectly legal to have a method being defined as "def myFunc" and returning a future. 

If you do not configure mypy, it demands that "await myFunc()" is "async myFunc()", but as written that is a false positive.

If you run mypy with default settings, you will have a lot of false positives. Our mypy settings are based on experience, including from bigger projects like HomeAssistant.

The doc is in general being updated, but it is not a fast process.

**Reply by MAKOMO** [reply]
Thanks for your comments. I will look into your mypy setting and see what I can learn.

**Reply by MAKOMO** [reply]
Still a statement like in line 20 of logging.py

```
log_file_name: str = None
```

seems wrong to me (even if accepted by mypy under some options) as one cannot trust its type annotation which states that `log_file_name` is (always) bound to a string.

**Reply by janiversen** [reply]
That is not correct, for a very long time, None have been implicit.

**Reply by MAKOMO** [reply]
But since end of 2020 PEP 484 suggest to make None explicit. See https://github.com/python/peps/pull/689

Types serve as documentation and to allow for automatic type checking. A simple `str` is here not as clear as it could be and keeping None implicit misses opportunities to discover unguarded access of operations on values of a type which fail if the value is None. In other languages (not Python) sloppy typing misses also opportunites for potential code optimizations during compilation.

**Reply by janiversen** [reply]
Let me put it like this, I am aware of a couple of projects, including this one that uses None implicit....but if you want a change then please make a pull request.

**Reply by MAKOMO** [reply]
I will try to come up with a pull request for this. One quick question: I see that in the current code the vertical bar is used for union types (eg. in transport/transport.py:get_protocol), a syntax introduced in Python 3.10. Similar `tuple` is used in type annotations which got introduced in Python 3.9. Is it ok to use this newer typing syntax (instead of the older `Tuple` and `Union`, `Optional`, ... alternatives) despite the target Python version for this lib is still v3.8?  I assume those type annotations are ignore anyhow by the Python on run-time and the CI runs mypy on newer Python release, correct?

**Reply by janiversen** [reply]
Any pull request need to pass CI, which runs linters on all versions, that is something we want to keep.

Changes more or less all declarations to be "x | None" hardly makes sense, it does not make the library more stable. However that is just my opinion as usual pull requests will be reviewed independent of my opinion.

**Reply by MAKOMO** [reply]
Ok

**Reply by alexrudd2** [reply]
(for other people reading this later)

Use `from __future__ import annotations` to permit `x | None` on Python3.8 (the earliest supported version)

**Comment by janiversen** [comment, marked as answer]
First part solved, zip now have content. https://github.com/pymodbus-dev/pymodbus/commit/63324a323da0b34e2a5d21bd18a12151cad5be51

Readthedocs is updated (dev).

**Comment by janiversen** [comment]
Solved examples as well:
https://github.com/pymodbus-dev/pymodbus/commit/64c182c7dfa9ffc86585794e87726d8a108d7b71

Readthedocs will pick it up (dev) in a couple of minutes.
